### ISSUE TITLE
dynamically  set tasks' replicas, with the range of [min, max]

### ISSUE THREAD

**Issue by umialpha** [volcano-sh/volcano#650]
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
> kind feature

**What happened**:
Suppose in Tensorflow area, when user submits a distributed Tensorflow job, he must decide the number of workers, and usually set MinAvalible := sum of replicas of tasks. 

**What you expected to happen**:
If we can set `replicas` to a range number, e.g. [min, max], we can enhance our scheduling ability.

1. If there are enough resources, we can set as many workers as possible ( <= max)
2. If there are less resources, we can start the job as fast as possible( >= min).
3. **If tensorflow workload (or any other workload) allow dynamic workers(e.g. auto-scaling), it gives kube-volcano more possibility to schedule.**

**Comment by umialpha** [comment]
As I know, there are some discussion in Tensorflow project about 'Dynamic workers'. 
[https://github.com/tensorflow/tensorflow/issues/11896](url)

**Comment by k82cn** [comment]
for now, we can set different value to `minAvailable` and `Replicas`: we'll make sure `minAvailable` pods will be started, and the other pods (`replicas - minAvailable`) will be scheduled by fair-share.

**Comment by umialpha** [comment]
> for now, we can set different value to `minAvailable` and `Replicas`: we'll make sure `minAvailable` pods will be started, and the other pods (`replicas - minAvailable`) will be scheduled by fair-share.

yeah, I know this functionality. But it has some gap with my disire. I need all workers starting at the same time, without any worker pending. What I want is that Volcano decide the best number of workers for me. What more, it can change the number dynamically during running time, But it need the job support which is a further topic. What do you think with this feature? I am writing some design doc with it, if you have interest, I will issue a PR for it.

**Comment by k82cn** [comment]
@umialpha , that makes sense to me, and I'd like to see the design document :)

**Comment by umialpha** [comment]
Hi @k82cn , I would like to discuss some thoughts about this feature.

1. If we set Replicas to the format like `[minReplicas, maxReplicas]` or add something like this , maybe we have to set each task to each PodGroup. We can't ensure each task meets the minReplicas request if all the tasks stay in the same PodGroup.

2. On which level should we bring in the feature, job controller level or scheduling level? The current workflow is that job controller creates the PodGroup and the Pods. Scheduler watches PodGroup events and Pods events, applying the scheduling policies on the Pods. Meanwhile the job controller watches the Pods status. In my opinion, this feature is one kind of scheduling features, it maybe better to stay on scheduling level. However, job controller controls the creation of Pods currently, How to bring this feature without big modifications on the workflow? My  first thought is let the controller creates `minReplicas` numbers or workers, while the scheduler decides the exact number of workers and notify the controller to do some extra things. I doubt whether it is the best practice.

**Comment by k82cn** [comment]
> Hi @k82cn , I would like to discuss some thoughts about this feature.
> 
> 1. If we set Replicas to the format like `[minReplicas, maxReplicas]` or add something like this , maybe we have to set each task to each PodGroup. We can't ensure each task meets the minReplicas request if all the tasks stay in the same PodGroup.

hm.... we used to consider support different replicas for different tasks; that's ok to support that in this feature. That's ok to have a google doc to identify the final solution :)

> 2. On which level should we bring in the feature, job controller level or scheduling level? The current workflow is that job controller creates the PodGroup and the Pods. Scheduler watches PodGroup events and Pods events, applying the scheduling policies on the Pods. Meanwhile the job controller watches the Pods status. In my opinion, this feature is one kind of scheduling features, it maybe better to stay on scheduling level. However, job controller controls the creation of Pods currently, How to bring this feature without big modifications on the workflow? My  first thought is let the controller creates `minReplicas` numbers or workers, while the scheduler decides the exact number of workers and notify the controller to do some extra things. I doubt whether it is the best practice.

IMO, the key part is scheduler :) Suggest to define interface/behavior of scheduler firstly; and then we can update job controller accordingly. That's ok to update job controller, as the overall workflow will be similar :)

**Comment by umialpha** [comment]
Hi @k82cn,
I find something interesting named "Enqueue" Action, focusing on delay pod creation. I think it may help with this feature. But I cannot find clues in volcano project. Has this feature been brought into the project? My thought is as follows:
1. Job Controller creates each PodGroup for each tasks.
2. PodGroup interacts with scheduler to decide how many pods can be created and reserve the resource for them.
3. Job Controller got notified and do some plugins(ssh, svc, etc.) functions.

**Comment by k82cn** [comment]
we already have enqueue action in volcano :) I'm thinking to enhance PodGroup to support minAvailable array/map to support different task, as we did not support gang-scheduling cross podgroup :).

**Comment by umialpha** [comment]
Hi @k82cn , correct me if I misunderstand something.
I think if we want to use `Enqueue` feature, we must put `equeue action` before `allocate action` because  `equeue action`  does some pre-filter functions. 

However, the design doc [https://github.com/volcano-sh/volcano/blob/master/docs/design/delay-pod-creation.md](url) says it is for delay pod creation. But I think the pods have been already created, actually they are not into the `allocate action`. 

I am not sure whether the doc is a litter misleading or I misunderstand something. I think the main entry point of the scheduler begins with the creation of pods. If I am right, maybe it is better to define the execution orders of actions internally.

**Comment by umialpha** [comment]
> we already have enqueue action in volcano :) I'm thinking to enhance PodGroup to support minAvailable array/map to support different task, as we did not support gang-scheduling cross podgroup :).

I think it is not hard to implement multi-gang scheduling feature. Here is my thought.

1. Modify `scheduler.api.JobInfo` structure.
- Change `PodGroup` field  to `map[string]*PodGroup` , the key is the name of the PodGroup.
-   Add a field named `MinAvailablePodGroup` with the type of `int` or `string[]`. `int` means the minimum number of PodGroups while `string[]` specifies the names of the `PodGroup`.
2. Modify JobInfo.Ready() function according to the new rule.


Does it make sense?

**Comment by k82cn** [comment]
> Modify JobInfo.Ready() function according to the new rule.

The `JobInfo` in scheduler is built from `PodGroup` instead of Volcano Job :)

**Comment by umialpha** [comment]
Let me clarify my thoughts more specifically.

We could broaden the JobInfo concept to the  group of PodGroups. By default, each PodGroup makes up each JobInfo, if we want to gang `PodGroup`, we have to specify them. let's say,

> apiVersion: scheduling.incubator.k8s.io/v1alpha1
> kind: PodGroup
> metadata:
>   name: fooGroup1
> spec:
>   minMember: 6
>   **groups: ["fooGroup2", "fooGroup3"]** 

Because we always want these three groups scheduled together, they can be treated as a whole JobInfo. When meeting `fooGroup1`, we first check whether `fooGroup2` or `fooGroup3` JobInfo exists, if any does, we add `fooGroup1` info with its tasks to this JobInfo, else we create a new JobInfo.

The other parts of kube-batch treat this JobInfo as usual, allocating the resources to the tasks, checking whether the `JobInfo` is ready, etc.

By broadening the concept of JobInfo instead of bringing a new structure, we could minimize our work.

**Comment by hzxuzhonghu** [comment]
The usability is the main concern. IMO, we donot change anything in PodGroup API. But we can support define multi podgroups cascading for a job. One parent with several subsets.

**Comment by Jeffwan** [comment]
@umialpha Tensorflow right now doesn't support dynamic training. Do you customize TF to support this or use different framework?

**Comment by umialpha** [comment]
> @umialpha Tensorflow right now doesn't support dynamic training. Do you customize TF to support this or use different framework?

Hi @Jeffwan , we are trying to customize TF by ourselves. There is some tricky workaround, e.g. reserve the resource, then restart the tf-job from the latest checkpoint. But that is not we want. Besides, volcano does not support `reserve plugins` right now.

Meanwhile, I find pytorch seemly support elastic scaling(https://github.com/pytorch/elastic). I am going to try it on Volcano. But to support elastic scaling feature, volcano has to support `ganging PodGroups` discussed above, which is what I am working on.

**Comment by umialpha** [comment]
> The usability is the main concern. IMO, we donot change anything in PodGroup API. But we can support define multi podgroups cascading for a job. One parent with several subsets.

Hi @hzxuzhonghu , we can make a structure to specify the groups of PodGroups without changing PodGroup API. What's your opinion about my thought of broadening the concept of `JobInfo`?

**Comment by Jeffwan** [comment]
> Hi @Jeffwan , we are trying to customize TF by ourselves. There is some tricky workaround, e.g. reserve the resource, then restart the tf-job from the latest checkpoint. But that is not we want. Besides, volcano does not support `reserve plugins` right now.

You can probably check https://github.com/sql-machine-learning/elasticdl This is a pluggable way to get gradients from training workers and then sync to new join members. 

> Meanwhile, I find pytorch seemly support elastic scaling(https://github.com/pytorch/elastic). I am going to try it on Volcano. But to support elastic scaling feature, volcano has to support `ganging PodGroups` discussed above, which is what I am working on.

Yes. I am actually working on this now, we partner with torch team to give better support elastic job support on EKS (it will work on other k8s distribution as well). I've finished naive way to deploy it on k8s, the second step is to have customized controller for it, and then resolve the scheduling and resource management problems.

**Comment by hzxuzhonghu** [comment]
>>The usability is the main concern. IMO, we donot change anything in PodGroup API. But we can support define multi podgroups cascading for a job. One parent with several subsets.

>Hi @hzxuzhonghu , we can make a structure to specify the groups of PodGroups without changing PodGroup API. What's your opinion about my thought of broadening the concept of JobInfo?

SGTM

**Comment by xiaogaozi** [comment]
As @Jeffwan mentioned, ElasticDL is a great solution for elastic scaling of TensorFlow job. But it doesn't support gang scheduling by now. I've create a issue https://github.com/sql-machine-learning/elasticdl/issues/1626 to ask the maintainer for feature request or integrate with Volcano.

**Comment by k82cn** [comment]
> Let me clarify my thoughts more specifically.
> 
> We could broaden the JobInfo concept to the group of PodGroups. By default, each PodGroup makes up each JobInfo, if we want to gang `PodGroup`, we have to specify them. let's say,
> 
> ```
>  apiVersion: scheduling.incubator.k8s.io/v1alpha1
>  kind: PodGroup
>  metadata:
>    name: fooGroup1
>  spec:
>    minMember: 6
>    groups: ["fooGroup2", "fooGroup3"] 
> ```
> 
> Because we always want these three groups scheduled together, they can be treated as a whole JobInfo. When meeting `fooGroup1`, we first check whether `fooGroup2` or `fooGroup3` JobInfo exists, if any does, we add `fooGroup1` info with its tasks to this JobInfo, else we create a new JobInfo.
> 
> The other parts of kube-batch treat this JobInfo as usual, allocating the resources to the tasks, checking whether the `JobInfo` is ready, etc.
> 
> By broadening the concept of JobInfo and limiting the change within the JobInfo concept instead of bringing a new structure, we could minimize our work.

I think we can add something like `PodGroupSpec` into `PodGroup`, e.g. 

 ```
  apiVersion: scheduling.incubator.k8s.io/v1alpha1
  kind: PodGroup
  metadata:
    name: fooGroup1
  spec:
    minMember: 6
    subGroups:
    - name: fooGroup2
       minMember: 3
     - name: fooGroup3
        minMember: 3  
```
That's also easier for other components to handle it, e.g. validation in admission , scheduler to build them.

**Comment by k82cn** [comment]
@xiaogaozi , @Jeffwan  is `subGroups` ok for you? If that's ok, I think we can start design doc for more detail.

**Comment by xiaogaozi** [comment]
May I ask, what case `subGroups` design for?

In our case, the TF job is running in parameter server mode (i.e. there're worker and PS pods). The worker pod could be stateless, but the PS pod will always stateful. So we could scale worker pods easily. Specifically we define the `min` and `max` number of worker pods (number of PS pods is fixed). Then Volcano decide how many worker pods could run for now.

**Comment by umialpha** [comment]
Hi @k82cn , I am actually working on this feature. But I start to doubt its necessity. I can set different priorities for different tasks, e.g. if I need 1 PS pod and [4, 10] workers, I can set high priority for PS task and low priority for worker task, in this way, I can ensure that if 5 pods are running, there must be 1 ps worker.
Could you please explain more about other user cases for this feature? 
BTW, I could send a PR for this feature design doc.

**Comment by umialpha** [comment]
> Yes. I am actually working on this now, we partner with torch team to give better support elastic job support on EKS (it will work on other k8s distribution as well). I've finished naive way to deploy it on k8s, the second step is to have customized controller for it, and then resolve the scheduling and resource management problems. Happy to collaborate with you.
@Jeffwan Do you have any open-source plan on your project?

**Comment by Jeffwan** [comment]
> @Jeffwan Do you have any open-source plan on your project?

Operator POC is almost done and I will submit the operator doc for review this week. For the scheduling challenges, I have not started yet. Watching this issue an considering good solutions now.

**Comment by habvt** [comment]
> > Meanwhile, I find pytorch seemly support elastic scaling(https://github.com/pytorch/elastic). I am going to try it on Volcano. But to support elastic scaling feature, volcano has to support `ganging PodGroups` discussed above, which is what I am working on.
> 
> Yes. I am actually working on this now, we partner with torch team to give better support elastic job support on EKS (it will work on other k8s distribution as well). I've finished naive way to deploy it on k8s, the second step is to have customized controller for it, and then resolve the scheduling and resource management problems. Happy to collaborate with you.

Hi all, I am looking for a way to use GPU sharing with torch elastic, and so any update on this?
